diff -up gnome-power-manager-2.32.0/man/Makefile.am.give-me-some-logs-please gnome-power-manager-2.32.0/man/Makefile.am
--- gnome-power-manager-2.32.0/man/Makefile.am.give-me-some-logs-please	2010-09-29 08:29:19.983923821 -0400
+++ gnome-power-manager-2.32.0/man/Makefile.am	2010-09-29 08:29:33.667924010 -0400
@@ -12,13 +12,13 @@ endif
 
 if HAVE_DOCBOOK2MAN
 gnome-power-manager.1: gnome-power-manager.sgml
-	docbook2man $? &> /dev/null
+	docbook2man $?
 
 gnome-power-preferences.1: gnome-power-preferences.sgml
-	docbook2man $? &> /dev/null
+	docbook2man $?
 
 gnome-power-statistics.1: gnome-power-statistics.sgml
-	docbook2man $? &> /dev/null
+	docbook2man $?
 endif
 
 clean-local :
