diff -Nru caja-open-terminal-0.2/src/Makefile.am caja-open-terminal-0.2.new/src/Makefile.am
--- caja-open-terminal-0.2/src/Makefile.am	2005-02-25 16:11:20.000000000 +0300
+++ caja-open-terminal-0.2.new/src/Makefile.am	2005-05-22 21:39:52.000000000 +0400
@@ -5,7 +5,6 @@
 	-I$(top_srcdir)					\
 	-I$(top_builddir)				\
 	$(WARN_CFLAGS)                                  \
-	-Werror                                         \
 	$(DISABLE_DEPRECATED_CFLAGS)			\
 	$(CAJA_CFLAGS)				\
 	$(MATECONF_CFLAGS)					\
diff -Nru caja-open-terminal-0.2/src/Makefile.in caja-open-terminal-0.2.new/src/Makefile.in
--- caja-open-terminal-0.2/src/Makefile.in	2005-03-04 17:16:23.000000000 +0300
+++ caja-open-terminal-0.2.new/src/Makefile.in	2005-05-22 21:40:04.000000000 +0400
@@ -186,7 +186,6 @@
 	-I$(top_srcdir)					\
 	-I$(top_builddir)				\
 	$(WARN_CFLAGS)                                  \
-	-Werror                                         \
 	$(DISABLE_DEPRECATED_CFLAGS)			\
 	$(CAJA_CFLAGS)				\
 	$(MATECONF_CFLAGS)					\

